### PR TITLE
Update breadcrumbs when workspace folders update

### LINKED ISSUE #113867: Breadcrumb supports expand all level by default
I want to search all symbol in current file, but Breadcrumb only show symbols that have expend already.
image

Compare to Jetbrains's IDE, it expend all level by default :)
image

This feature is convenient for us to quick search symbol

### LINKED ISSUE #154615: Breadcrumbs don't update when workspace folders are updated
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version:Version: 1.70.0-insider (user setup)
Commit: 3104db414c8fedcf6e4493f14da7df0b7413853a
Date: 2022-07-08T05:20:13.606Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Windows_NT x64 10.0.19044
- OS Version:

Steps to Reproduce:

1. File>Add folder to workspace (folder a), File>Add folder to workspace (folder a/b)
2. Open file a/b/c.json, breadcrumbs show "b > c.json"
3. Right click folder b and choose "Remove Folder From Workspace"
4. Breadcrumb shows "b > c.json" (expected: "a > b > c.json" )
5. Close and reopen c.json, breadcrumbs now show correctly

Also:

1. "Workspaces: Open Workspace Configuration File"
2. Arrange c.json on the left and the workspace configuration file on the right
3. Add `"name": "foo"` to `folders[0]` and save the workspace configuration file
4. Breadcrumb shows "a > b > c.json" (expected: "foo > b > c.json" )

### PR DESCRIPTION
**Pull request by r3m0t**
This PR fixes #154615

Test plan at #154615

### CODE REVIEW

**Review by jrieken — CHANGES_REQUESTED** [review]
The change overall looks good but it should be done not inn the view but in the model, e.g [`BreadcrumbsModel`](https://github.com/microsoft/vscode/blob/d1de1c651cd08c807de91f927fdd3ffa03f2b97f/src/vs/workbench/browser/parts/editor/breadcrumbsModel.ts#L37). It should listen to the workspace change and trigger its change event

**Comment by r3m0t** [comment]
@jrieken can you have a look before I start looking at the tests?

**Review by jrieken — DISMISSED** [review]
Changing for breadcrumbs looks good. Assigning @bpasero for changes in the labels-world

**Review by bpasero — CHANGES_REQUESTED** [review]
Not liking the changes in `labels.ts`, I find `recomputeOnWorkspaceFolderChange` strange. Can we not somehow figure out if an update is needed when a transition happens into workspace? Isn't that like the only time we want to update?

File: src/vs/workbench/browser/labels.ts [hunk 1]
```diff
@@ -362,6 +372,15 @@ class ResourceLabelWidget extends IconLabel {
 		this.render({ updateIcon: true, updateDecoration: false });
 	}
 
+	notifyWorkspaceFoldersChanges(): void {
+		if (this.recomputeOnWorkspaceFolderChange) {
```
[the remark this answers is not in the document]
Comment (bpasero) [reply src/vs/workbench/browser/labels.ts:374]: I would prefer the "not messy" solution.

File: src/vs/workbench/browser/labels.ts [hunk 2]
```diff
@@ -444,6 +463,7 @@ class ResourceLabelWidget extends IconLabel {
 
 		this.label = label;
 		this.options = options;
+		this.recomputeOnWorkspaceFolderChange = recomputeOnWorkspaceFolderChange;
```
Comment (bpasero) [inline src/vs/workbench/browser/labels.ts:466]: This seems strange to me.

**Comment by r3m0t** [comment]
@bpasero the case is that the root folder name changes in the workspace, which can happen by editing the workspace file.

**Review by bpasero — CHANGES_REQUESTED** [review]
Thanks, but I am still not so happy: now we seem to be overwriting the name of a label to be the workspace folder name, but shouldnt that be something only the user of the label set? Isnt the label service typically the driver of a label name?

My worry is that someone sets a name to a label and now gets overwritten by the name of the workspace folder.

What is the actual issue to solve here? Can we not listen in breadcrumbs and update when workspaces change?

**Comment by r3m0t** [comment]
Well, name can be set in three ways - consumer of label service calls `setFile` and consumer of label service calls `setResource`. (The third way is `setLabel` but it isn't relevant)

In this case, "consumer of labels" is "breadcrumbs" and it calls `setFile`.

The best encapsulation is, the state (`ResourceLabelWidget.label.name`) is maintained by whoever set it in the first place. So if `setFile` is used, label service set initial label and should maintain it to the correct value. If `setResource` is used, consumer of label service set initial label by passing it as the argument `IResourceLabelProps.name` and it will have to maintain it to the correct value as well by later calls to `setResource`.

Without this encapsulation, basically `setFile` is no longer convenient because it doesn't remove the burden on the consumer of label service to be interested in workspace names.

Here's the code from breadcrumbs:

* When items change, eg a different tab is selected, don't re-render the common prefix-

https://github.com/microsoft/vscode/blob/764d22738fcdeaa7b02b384f2d8e67f47e47d115/src/vs/base/browser/ui/breadcrumbs/breadcrumbsWidget.ts#L278

Equality of each breadcrumb for calculating the common prefix is purely by URI - 

https://github.com/microsoft/vscode/blob/764d22738fcdeaa7b02b384f2d8e67f47e47d115/src/vs/workbench/browser/parts/editor/breadcrumbsControl.ts#L125

Rendering the new breadcrumbs is by `setFile` -

https://github.com/microsoft/vscode/blob/764d22738fcdeaa7b02b384f2d8e67f47e47d115/src/vs/workbench/browser/parts/editor/breadcrumbsControl.ts#L134

So if label service doesn't listen to `onDidChangeWorkspaceFolders`, then we'll need to change the `equals` implementation to consider the workspace folder name and change the `render` implementation to call `setResource` `if FileElement.kind === FileKind.RootFolder` so that it can pass the name to the label widget. And, other users of the label service that call `setFile` could need to put in similar work!

So, let's continue with the label service maintaining the state `ResourceLabelWidget.label.name` and let consumers continue to use `setFile` and not look at the workspace folder names.

To prevent accidental updates from the `onDidChangeWorkspaceFolders`, there's two options, one is to save a flag in `setFile`. The other is to look at `IWorkspaceFoldersChangeEvent`? Like, `if IWorkspaceFoldersChangeEvent.changed[i].uri == roResource(this.label) this.setFile()` . But, the event doesn't give the old folder name so there's no way to check if the label already matches the old root folder name.

I still prefer the original version as it expresses the intent most clearly. Label service set `name` therefore it should update it. https://github.com/microsoft/vscode/pull/154616/files/4b241555ad2aa48adb695955efda856b397ace75

**Comment by r3m0t** [comment]
Thanks @bpasero , did you mean to check the value of computedWorkspaceFolderLabel  in notifyWorkspaceFoldersChanges ? In your current version, it looks like you're only checking the type. It also needs to be added to clear() method. I will get to this in the next few days.

**Comment by bpasero** [comment]
Yeah feel free to further tweak it, thanks!

**Comment by r3m0t** [comment]
@bpasero @jrieken all ready

**Review by bpasero — APPROVED** [review]

**Pull request merged**